glide-core/src/socket_listener.rs

```diff
@@ -924,23 +806,109 @@ pub fn start_socket_listener_internal<InitCallback>(
+    }
+
+    // Retry with write lock, will be dropped upon the function completion
+    let mut sockets_write_guard = INITIALIZED_SOCKETS
+        .write()
+        .expect("Failed to acquire sockets db write guard");
+    if sockets_write_guard.contains(&socket_path) {
+        init_callback(Ok(socket_path.clone()));
+        return;
+    }
+
+    let (tx, rx) = std::sync::mpsc::channel();
+    let socket_path_cloned = socket_path.clone();
+    let init_callback_cloned = init_callback.clone();
+    let tx_cloned = tx.clone();
     thread::Builder::new()
         .name("socket_listener_thread".to_string())
         .spawn(move || {
-            let runtime = Builder::new_current_thread().enable_all().build();
-            match runtime {
-                Ok(runtime) => {
-                    let mut listener = Disposable::new(SocketListener::new(
-                        socket_path.unwrap_or_else(get_socket_path),
-                    ));
-                    runtime.block_on(listener.listen_on_socket(init_callback));
+            let init_result = {
+                let runtime = Builder::new_current_thread().enable_all().build();
+                if let Err(err) = runtime {
+                    log_error(
+                        "listen_on_socket",
+                        format!("Error failed to create a new tokio thread: {err}"),
+                    );
+                    return Err(err);
                 }
-                Err(err) => init_callback(Err(err.to_string())),
+
+                runtime.unwrap().block_on(async move {
+                    let listener_socket = UnixListener::bind(socket_path_cloned.clone());
+                    if let Err(err) = listener_socket {
+                        log_error(
+                            "listen_on_socket",
+                            format!("Error failed to bind listening socket: {err}"),
+                        );
+                        return Err(err);
+                    }
+                    let listener_socket = listener_socket.unwrap();
+
+                    // signal initialization success
+                    init_callback(Ok(socket_path_cloned.clone()));
+                    let _ = tx.send(true);
+
+                    let local_set_pool = LocalPoolHandle::new(num_cpus::get());
+                    loop {
+                        match listener_socket.accept().await {
+                            Ok((stream, _addr)) => {
+                                local_set_pool
+                                    .spawn_pinned(move || listen_on_client_stream(stream));
+                            }
+                            Err(err) => {
+                                log_error(
+                                    "listen_on_socket",
+                                    format!("Error accepting connection: {err}"),
+                                );
+                                break;
+                            }
+                        }
+                    }
```

Comment:
You should consider making this method returning a `Result`.

This highlighted lines above can then become:

```rust
let stream = listener_socket.accept().await?.0;
local_set_pool.spawn_pinned(move || listen_on_client_stream(stream));
```

Reply:
isnt having an await? will make it impossible to log "error accepting connection"

Reply:
I assume that you meant here that we will lose the context on which the error occurred (e.g. "accept error"). We can use `map_err` for this OR we can use a more elegant way by creating a our own set of errors:

For example, lets say we have defined a new result: `MyResult`


```rust
use thiserror::Error;

#[derive(Error, Debug)]
pub enum MyError {
    #[error("Accept error. {0}")]
    AcceptError(String),
}

// we can not use this
listener_socket.accept()
    .await
    .map_err(|e| MyError::AcceptError(e.to_string()))?
    .0;
```

Reply:
really all this for what?